src/tests/integration/D3MCompound.t.sol

```diff
@@ -0,0 +1,1234 @@
+        assertEq(cDai.borrow(amountToBorrow), 0);
+
+        // MCD shutdowns
+        end.cage();
+        end.cage(ilk);
+
+        hevm.warp(block.timestamp + end.wait());
+
+        // Force remove all the dai from vow so it can call end.thaw()
+        hevm.store(
+            address(vat),
+            keccak256(abi.encode(address(vow), uint256(5))),
+            bytes32(0)
+        );
+
+        end.thaw();
+
+        assertRevert(address(d3mHub), abi.encodeWithSignature("exec(bytes32)", ilk), "D3MHub/end-debt-already-set");
+    }
+
+    function test_unwind_culled_then_mcd_caged() public {
+        uint256 currentLiquidity = dai.balanceOf(address(cDai));
+
+        // Lower by 50%
+        uint256 targetBorrowRate = _setRelBorrowTarget(5000);
+        assertEqInterest(getBorrowRate(), targetBorrowRate);
+
+        (uint256 pink, uint256 part) = vat.urns(ilk, address(d3mCompoundPool));
+        assertGt(pink, 0);
+        assertGt(part, 0);
+
+        // Someone else borrows
+        // Someone else borrows
+        uint256 amountSupplied = cDai.balanceOfUnderlying(address(d3mCompoundPool));
+        uint256 amountToBorrow = currentLiquidity + amountSupplied / 2;
+        assertEq(cDai.borrow(amountToBorrow), 0);
+
+        d3mHub.cage(ilk);
+
+        (, , uint256 tau, , ) = d3mHub.ilks(ilk);
+
+        hevm.warp(block.timestamp + tau);
+        hevm.roll(block.number + tau / 15);
+
+        uint256 daiEarned = cDai.balanceOfUnderlying(address(d3mCompoundPool)) - pink;
+
+        VowLike(vow).heal(
+            _min(
+                vat.sin(vow) - VowLike(vow).Sin() - VowLike(vow).Ash(),
+                vat.dai(vow)
+            )
+        );
+        uint256 originalSin = vat.sin(vow);
+        uint256 originalDai = vat.dai(vow);
+        // If the whole Sin queue would be cleant by someone,
+        // originalSin should be 0 as there is more profit than debt registered
+        assertGt(originalDai, originalSin);
+        assertGt(originalSin, 0);
+
+        d3mHub.cull(ilk);
+
+        // After cull, the debt of the position is converted to bad debt
+        assertEq(vat.sin(vow), originalSin + part * RAY);
+
+        // CDP grabbed and ink moved as free collateral to the deposit contract
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(d3mCompoundPool));
+        assertEq(ink, 0);
+        assertEq(art, 0);
+        assertEq(vat.gem(ilk, address(d3mCompoundPool)), pink);
+        assertGe(cDai.balanceOfUnderlying(address(d3mCompoundPool)), pink);
+
+        // MCD shutdowns
+        originalDai = originalDai + vat.dai(VowLike(vow).flapper());
+        end.cage();
+
+        if (originalSin + part * RAY >= originalDai) {
+            assertEq(vat.sin(vow), originalSin + part * RAY - originalDai);
+            assertEq(vat.dai(vow), 0);
+        } else {
+            assertEq(vat.dai(vow), originalDai - originalSin - part * RAY);
+            assertEq(vat.sin(vow), 0);
+        }
+
```

Comment:
We could add an `assertRevert` here to show that `end.cage(ilk)` won't work until `uncull(ilk)` happens

Reply:
Handled in `eba23f7018`